blog/2016-04-20-futhark-and-pygame.rst

```diff
@@ -0,0 +1,215 @@
+the simulation window as expected.
+
+
+The Three Other Ones
+--------------------
+
+In the fluid simulator you can add both particles and forces.  See for yourself:
+
+.. raw:: html
+
+   <video src="/static/fluid-2016.04.20.webm"
+          poster="/static/fluid-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/fluid-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/fluid-2016.04.20.webm">fluid-2016.04.20.webm</a></p>
+
+My GPU is not the best one around, so I am running this in a fairly
+small window.
+
+The Mandelbrot Explorer is also pretty nifty:
+
+.. raw:: html
+
+   <video src="/static/mandelbrot-2016.04.20.webm"
+          poster="/static/mandelbrot-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/mandelbrot-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/mandelbrot-2016.04.20.webm">mandelbrot-2016.04.20.webm</a></p>
+
+In the end of the video, I switch to NumPy's internal Mandelbrot generator.
+
+Finally, there is the HotSpot 2D Heat Equation GUI.  You can see its
+performance `here </performance.html#hotspot-futhark-rodinia>`_.  This
+visualisation is pretty silly.
+
+.. raw:: html
+
+   <video src="/static/hotspot-2016.04.20.webm"
+          poster="/static/hotspot-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/hotspot-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/hotspot-2016.04.20.webm">hotspot-2016.04.20.webm</a></p>
+
+
+Use Cases
+---------
+
+Futhark is an optimising compiler which takes an *entire program* as
+input.  As such, its optimisations are not directed at separate
+functions, but rather the program as a whole.  This is important to
+keep in mind when developing Futhark-PyGame programs, since it means
+that we would like to have as few calls as possible to Futhark from
+Python, and keep as much code as possible inside Futhark.
+
+This is in stark contrast to how computing libraries, e.g. NumPy,
+usually work.  They consist of many primitive functions, and expect
+the programmer to structure them together using the host language, in
+this case Python.
+
+In conclusion, The Futhark-PyGame combo is best when every step of the
+visualisation -- i.e. every call to Futhark -- is compute-intensive.
```

Comment:
This is true.  But still, it works surprisingly well already, and will work even better in the future when we implement lazy on-demand motion of data between CPU and GPU.